dizoo/gym_anytrading/envs/stocks_env.py

```diff
@@ -0,0 +1,109 @@
+from typing import Any
+import numpy as np
+from dizoo.gym_anytrading.envs.trading_env import TradingEnv, Actions, Positions
+from ding.utils import ENV_REGISTRY
+from ding.torch_utils import to_ndarray
+
+
+@ENV_REGISTRY.register('stocks-v0')
+class StocksEnv(TradingEnv):
+
+    def __init__(self, cfg):
+
+        super().__init__(cfg)
+
+        self.trade_fee_bid_percent = 0.01  # unit
+        self.trade_fee_ask_percent = 0.005  # unit
+
+    def _process_data(self, start_idx: int = None) -> Any:
+        '''
+        Overview:
+            used by env.reset(), process the raw data.
+        Arguments:
+            - start_idx (int): the start tick; if None, then randomly select.
+        Returns:
+            - prices: the close.
+            - signal_features: feature map
+        '''
+
+        # ====== build feature map ========
+        all_feature_name = ['Close', 'Open', 'High', 'Low', 'Adj Close', 'Volume']
+        all_feature = {k: self.df.loc[:, k].to_numpy() for k in all_feature_name}
+        # add feature "Diff"
+        prices = self.df.loc[:, 'Close'].to_numpy()
+        diff = np.insert(np.diff(prices), 0, 0)
+        all_feature_name.append('Diff')
+        all_feature['Diff'] = diff
+        # =================================
+
+        # you can select features you want
+        selected_feature_name = ['Close', 'Diff', 'Volume']
+        selected_feature = np.column_stack([all_feature[k] for k in selected_feature_name])
+
+        # validate index
+        if start_idx is None:
+            self.start_idx = np.random.randint(self.window_size, len(self.df) - self._cfg.eps_length - 1)
+        else:
+            self.start_idx = start_idx
+
+        self._start_tick = self.start_idx
+        self._end_tick = self._start_tick + self._cfg.eps_length - 1
+
+        return prices, selected_feature
+
+    def _calculate_reward(self, action: int) -> np.float32:
+        step_reward = 0.
+        current_price = (self.raw_prices[self._current_tick])
+        last_trade_price = (self.raw_prices[self._last_trade_tick])
+        ratio = current_price / last_trade_price
+        cost = np.log((1 - self.trade_fee_ask_percent) * (1 - self.trade_fee_bid_percent))
+
+        if action == Actions.BUY.value and self._position == Positions.SHORT:
+            step_reward = np.log(2 - ratio) + cost
+
+        if action == Actions.SELL.value and self._position == Positions.LONG:
+            step_reward = np.log(ratio) + cost
+
+        if action == Actions.DOUBLE_SELL.value and self._position == Positions.LONG:
+            step_reward = np.log(ratio) + cost
+
+        if action == Actions.DOUBLE_BUY.value and self._position == Positions.SHORT:
+            step_reward = np.log(2 - ratio) + cost
+
+        step_reward = to_ndarray([step_reward]).astype(np.float32)
+
+        return step_reward
+
+    def max_possible_profit(self) -> float:
```

Comment:
[No need to change, just a quick discussion]

This function is fine with no transaction fees, as it takes full advantage of each stock price up and down
But when considering transaction fees, I will show you an example:
Given the transaction fee is `1$` on both long and short: 
T0: the price goes up `0.8$`;
T1: the price goes down `0.2$`;
T2: the price goes up `0.7$`.
According to the function, you should not buy or sell in this process.
However, it is possible to earn `0.8 - 0.2 + 0.7 - 1 = 0.3$`.

I just found the drawback but no perfect solution at hand, we can discuss this later.

Reply:
Actually, the calculation is based on ratio.
The calculation formula of profit is $\prod_{buying\ long}(r_{curr}/r_{pre}\ *\ cost) + \prod_{short\ selling}((2-r_{curr}/r_{pre})\ *\ cost)$. The profit is initialized to _1_ , which means 100%.
Take buying long as an example, the code shows that
> tmp_profit = profit * (current_price / last_trade_price) * (1 - self.trade_fee_ask_percent) * (1 - self.trade_fee_bid_percent)

T0: close is 212.557556;
T1: close is 215.340347;

transaction fee ratio is (1 - 0.01) * (1 - 0.005) = 0.98505
then 

T1_profit = T0_profit * (215.340347 / 212.557556) * 0.98505 
=  T0_profit * 1.013092 * 0.98505 
= T0_profit * 0.997946

In the case, although the closing price is rising, we should not take "buying long". 
And the code as following avoids this situation.
>  profit = max(profit, tmp_profit)

Reply:
Oh! I know what you mean. I'll reconsider it.

Reply:
This is a lower bound of _max_possible_profit_. Maybe we need to use dynamic programming to accurately calculate it.